### PR TITLE
Entanglement entropy of pure bipartite states

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**

[Quantifying the entanglement](https://www.quantiki.org/wiki/entanglement-measure) for example by determining the [entropy of entanglement](https://en.wikipedia.org/wiki/Entropy_of_entanglement) of bipartite quantum states can be useful for applications of quantum information (e.g., in quantum cryptography).

One way of determining entanglement entropy is to compute the [von Neumann entropy](https://docs.pennylane.ai/en/stable/code/api/pennylane.qinfo.transforms.vn_entropy.html) of one of the reduced states (see [reduced density matrix feature](https://docs.pennylane.ai/en/stable/code/api/pennylane.qinfo.transforms.reduced_dm.html) here).

It could be useful to add a user interface for such a measure specifically for the bipartite quantum state case.

**Description of the Change:**

The von Neumann entanglement entropy feature is now added to ```math``` and ```qinfo```. ```qml.math.vn_entanglement_entropy()``` returns the Von Neumann entanglement entropy between two subsystems in a given density matrix. ```qml.qinfo.vn_entanglement_entropy()``` computes the Von Neumann entanglement entropy from a ```QNode``` returning a ```state()```.

**Benefits:**

It enables users to find the entanglement entropy of a pure bipartite state.

**Possible Drawbacks:**

The von Neumann entanglement entropy is only defined for pure bipartite states. The current implementation does not perform any checks on if the overall input state is pure or mixed, since such a check would be jax.jit-unfriendly.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/3150

**Related Shortcut Story**
[sc-55206]

### CODE REVIEW

File: pennylane/math/quantum.py [hunk 1]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
     return vn_entropy_1 + vn_entropy_2 - vn_entropy_12
 
 
+# pylint: disable=too-many-arguments
+def vn_entanglement_entropy(
+    state, indices0, indices1, base=None, check_state=False, c_dtype="complex128"
+):
+    r"""Compute the Von Neumann entanglement entropy between two subsystems in a given state
```
Comment (PietropaoloFrisoni) [inline pennylane/math/quantum.py:740]:
```suggestion
    r"""Compute the Von Neumann entanglement entropy between two subsystems in a given state.

```

File: pennylane/math/quantum.py [hunk 2]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+    state, indices0, indices1, base=None, check_state=False, c_dtype="complex128"
+):
+    r"""Compute the Von Neumann entanglement entropy between two subsystems in a given state
+
+    .. math::
+
+        S(\rho_A) = -\text{Tr}[\rho_A log \rho_A] = -\text{Tr}[\rho_B log \rho_B] = S(\rho_B)
+
+    where :math:`S` is the von Neumann entropy, and :math:`\rho_A = \text{Tr_B}[\rho_{AB}]` and
+    :math:`\rho_B = \text{Tr_A}[\rho_{AB}]` are the reduced density matrices for each partition.
```
Comment (PietropaoloFrisoni) [inline pennylane/math/quantum.py:747]: I don't know if this was intentional, but perhaps you meant to write `\text{Tr}_A` instead of `\text{Tr_A}` etc.

File: pennylane/math/quantum.py [hunk 3]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+        float: The von Neumann entanglement entropy of the bipartite state.
+
+    **Examples**
+
+    The entanglement entropy between subsystems for a state vector can be returned as follows:
+    >>> x = np.array([0, -1, 1, 0]) / np.sqrt(2)
+    >>> qml.math.vn_entanglement_entropy(x, indices0=[0], indices1=[1])
+    0.6931471805599453
+
+    It is also possible to change the log basis.
```
Comment (PietropaoloFrisoni) [inline pennylane/math/quantum.py:775]:
```suggestion
    It is also possible to change the logarithm base:

```

File: pennylane/qinfo/transforms.py [hunk 4]
```diff
@@ -412,6 +422,42 @@ def _mutual_info_qnode(self, qnode, targs, tkwargs):
...
+    two subsystems constituting a pure bipartite quantum state. The entropy of entanglement is the
+    Von Neumann entropy of the reduced density matrix for any of the subsystems. If it is non-zero,
+    it indicates the two subsystems are entangled.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): A quantum circuit returning a :func:`~pennylane.state`.
+        wires0 (Sequence(int)): List of wires in the first subsystem.
+        wires1 (Sequence(int)): List of wires in the second subsystem.
+        base (float): Base for the logarithm. If None, the natural logarithm is used.
+
```
Comment (PietropaoloFrisoni) [inline pennylane/qinfo/transforms.py:448]:
```suggestion
**kwargs: Additional keyword arguments.
```

File: pennylane/qinfo/transforms.py [hunk 5]
```diff
@@ -412,6 +422,42 @@ def _mutual_info_qnode(self, qnode, targs, tkwargs):
...
+    tape, wires0: Sequence[int], wires1: Sequence[int], base: float = None, **kwargs
+):
+    r"""Compute the Von Neumann entanglement entropy from a circuit returning a :func:`~pennylane.state`:
+
+    .. math::
+
+        S(\rho_A) = -\text{Tr}[\rho_A log \rho_A] = -\text{Tr}[\rho_B log \rho_B] = S(\rho_B)
+
+    where :math:`S` is the von Neumann entropy; :math:`\rho_A = \text{Tr_B}[\rho_{AB}]` and
+    :math:`\rho_B = \text{Tr_A}[\rho_{AB}]` are the reduced density matrices for each partition.
```
Comment (PietropaoloFrisoni) [inline pennylane/qinfo/transforms.py:436]: Same as above

File: pennylane/math/quantum.py [hunk 6]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+    The Von Neumann entanglement entropy is a measure of the degree of quantum entanglement between
+    two subsystems constituting a pure bipartite quantum state. The entropy of entanglement is the
+    Von Neumann entropy of the reduced density matrix for any of the subsystems. If it is non-zero,
+    it indicates the two subsystems are entangled.
+
+    Each state must be given as a density matrix. To find the mutual information given
+    a pure state, call :func:`~.math.dm_from_state_vector` first.
+
+    Args:
+        state (tensor_like): ``(2**N)`` state vector or ``(2**N, 2**N)`` density matrix.
```
Comment (trbromley) [inline pennylane/math/quantum.py:758]: I'm confused about statevector vs dm support as these statements are conflicting.

File: pennylane/math/quantum.py [hunk 7]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+        indices0 (list[int]): List of indices in the first subsystem.
+        indices1 (list[int]): List of indices in the second subsystem.
+        base (float): Base for the logarithm. If None, the natural logarithm is used.
```
Comment (trbromley) [inline pennylane/math/quantum.py:761]:
```suggestion
        base (float): Base for the logarithm. If ``None``, the natural logarithm is used.

```

File: pennylane/math/quantum.py [hunk 8]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+        check_state (bool): If True, the function will check the state validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.
```
Comment (trbromley) [inline pennylane/math/quantum.py:763]: Why do we need to pass this rather than have it inferred from `state`?

File: pennylane/math/quantum.py [hunk 9]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+
+    Returns:
+        float: The von Neumann entanglement entropy of the bipartite state.
+
+    **Examples**
+
+    The entanglement entropy between subsystems for a state vector can be returned as follows:
+    >>> x = np.array([0, -1, 1, 0]) / np.sqrt(2)
+    >>> qml.math.vn_entanglement_entropy(x, indices0=[0], indices1=[1])
+    0.6931471805599453
```
Comment (trbromley) [inline pennylane/math/quantum.py:773]:
```suggestion
    The entanglement entropy between subsystems for a state vector can be returned as follows:

    >>> x = np.array([0, -1, 1, 0]) / np.sqrt(2)
    >>> qml.math.vn_entanglement_entropy(x, indices0=[0], indices1=[1])
    0.6931471805599453
```

File: pennylane/math/quantum.py [hunk 10]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+
+    It is also possible to change the logarithm base:
```
Comment (trbromley) [inline pennylane/math/quantum.py:775]:
```suggestion
    It is also possible to change the logarithm base:

```

File: pennylane/math/quantum.py [hunk 11]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+    >>> qml.math.vn_entanglement_entropy(x, indices0=[0], indices1=[1], base=2)
+    1
+
+    Similarly, the quantum state can be provided as a density matrix:
```
Comment (trbromley) [inline pennylane/math/quantum.py:779]:
```suggestion
    Similarly, the quantum state can be provided as a density matrix:

```

File: pennylane/math/quantum.py (hunk 9 above)
Comment (trbromley) [reply pennylane/math/quantum.py:773]: https://xanaduai-pennylane--5306.com.readthedocs.build/en/5306/code/api/pennylane.math.vn_entanglement_entropy.html

File: pennylane/math/quantum.py [hunk 12]
```diff
@@ -733,6 +733,80 @@ def _compute_mutual_info(
...
+    Von Neumann entropy of the reduced density matrix for any of the subsystems. If it is non-zero,
+    it indicates the two subsystems are entangled.
+
+    Each state must be given as a density matrix. To find the mutual information given
+    a pure state, call :func:`~.math.dm_from_state_vector` first.
+
+    Args:
+        state (tensor_like): ``(2**N)`` state vector or ``(2**N, 2**N)`` density matrix.
+        indices0 (list[int]): List of indices in the first subsystem.
+        indices1 (list[int]): List of indices in the second subsystem.
```
Comment (trbromley) [inline pennylane/math/quantum.py:760]: Should `indices0 | indices1 = {0, 1, 2, ..., N}`?

File: pennylane/math/quantum.py (hunk 6 above)
Comment (astralcai) [reply pennylane/math/quantum.py:758]: Ooops, I'll update the arg description.

File: pennylane/math/quantum.py (hunk 8 above)
Comment (astralcai) [reply pennylane/math/quantum.py:763]: I am not sure, I used the same arguments as in `mutual_info`

**Review by PietropaoloFrisoni — APPROVED** [review]
👍👍👍

File: pennylane/math/quantum.py (hunk 8 above)
Comment (trbromley) [reply pennylane/math/quantum.py:763]: No problem, we can check in a Core standup.

Comment (astralcai) [reply pennylane/math/quantum.py:763]: According to Romain, this is required by the old device API where it is possible to set the dtype on the device, so this was needed to match the device dtype.

Comment (trbromley) [reply pennylane/math/quantum.py:763]: Thanks for checking! It's still not clear to me why that can't be inferred by the `state` :thinking: Can the `state` be a different dtype to what the device wants?

Comment (astralcai) [reply pennylane/math/quantum.py:763]: I tried removing all `c_dtype` casting from the module and it didn't seem to cause any test failures, so I'm unsure if this is still necessary anymore. Although I don't think it's best to make such a change in this PR. For now I think it's reasonable to make the call signature of `vn_entanglement_entroy` consistent with all other math functions.

File: pennylane/math/quantum.py (hunk 12 above)
Comment (astralcai) [reply pennylane/math/quantum.py:760]: Not necessarily, the two subsystems don't have to cover the fully system.

Comment (trbromley) [reply pennylane/math/quantum.py:760]: Great! What prompts my question is that it's not clear in that page what `indices0` or `indices1` should in principle be.

Comment (astralcai) [reply pennylane/math/quantum.py:760]: I would say that they are indices of the wires that constitutes each sub system, not the wire labels, but the indices of the wires.

Comment (trbromley) [reply pennylane/math/quantum.py:760]: Agreed, we just never say that for $n$ wires we expect our indices to be subsets of $\\{1, \ldots, n \\}$

**Review by mudit2812 — APPROVED** [review]
